pkg/webreg/graphviz.go

```diff
@@ -2,183 +2,224 @@ package webreg
+	// label is rendered as plain text.
 	linkable bool
 }
 
+// subgraph is a group of sequential steps and/or other subgraphs
+// Used to render both registry chains and the pre/test/post steps of a
+// workflow.
 type subgraph struct {
-	label       string
-	nodes       []int
-	subgraphs   []int
-	firstNode   int
-	lastNode    int
-	hasSGParent bool
-	linkable    bool
+	label     string
+	nodes     []int
+	subgraphs []int
 }
 
+// edge connects node objects in the final drawing
+// Every edge connects two nodes in the final output.  If an edge crosses one
+// or more chain boundaries, either on the incoming or outgoing direction, it
+// is clipped to the outermost box used to render the chains.  In this case,
+// one or both of the `type` members are set to `subgraphType` and the
+// corresponding `graph` members are set to the sub-graph against which the
+// edge is to be clipped (options `lhead` and `ltail` in `graphviz`).
 type edge struct {
-	srcType edgeType
-	dstType edgeType
-	src     int
-	dst     int
+	srcType, dstType   edgeType
+	srcNode, dstNode   int
+	srcGraph, dstGraph int
 }
 
+// graph is the container for all elements
+// All other edge, sub-graph, etc. references are indices into this structure.
 type graph struct {
 	label     string
 	nodes     []node
-	subgraphs []subgraph
 	edges     []edge
+	subgraphs []subgraph
+}
+
+// graphBuilder maintains state required to build a `graph` object
+type graphBuilder struct {
+	chains registry.ChainByName
+	// The final graph.
+	// Can be initialized prior to the construction (e.g. with a label).
+	graph graph
+	// The next edge to be emitted when a new node is found.
+	//
+	// This variable is the only state necessary to create edges between nodes
+	// of the graph.  It is modified throughout the traversal of the various
+	// objects so that it contains the required information to emit a new edge
+	// when a node is created, namely:
+	//
+	// - Its initial value is `noEdge`, so that the first node does not get an
+	//   incoming edge.
+	// - Whenever the traversal enters a chain, `dstType` is set to
+	//   `subgraphType`, so that the incoming edge is clipped against the
+	//   chain's bounding box.
+	// - Whenever the traversal leaves a chain, `srcType` is similarly set to
+	//   `subgraphType` for the same reason, for the outgoing edge.
+	// - In both of the previous cases, `dstGraph`/`srcGraph` is set
+	//   accordingly.  It contains the index of the chain's sub-graph.  Note
+	//   that this is done by `addSubgraph` at every level and always at the
+	//   end of the function, so that the edge is clipped against the outermost
+	//   chain.
+	// - Whenever a node (i.e. a step) is added to the graph, an edge is added
+	//   based on the current state of the variable.  Its destination is the
+	//   newly added node.  This variable is then changed in preparation for
+	//   the subsequent edge (its source is the one just added).
+	edge edge
```

Comment:
The description is outstanding. This is just my personal taste and nothing else: I've noticed `addSubgraph` is recursive and it manipulates `edge` across its calls, it's fine even though I find it easier to follow recursive algorithms that have the state as a parameter rather than a member of a class.
No need to change anything by the way.

Reply:
> E.g. the edge between `ipi-install-install` and the `test` sub-graph, from the same `step-to-chain` as before:

To illustrate:

```
addSubgraph - pre
    addNode - ipi-deprovision-deprovision
    addSubgraph - ipi-install
        addNode - ipi-install-rbac
        addNode - ipi-install-install
            [edge is set]
        [edge is modified (to make `ipi-install` its source)]
    [edge is modified (to make `pre` its source)]
addSubgraph - test
    addNode - empty
        [the actual edge in the graph is created]
    [edge is modified (to make `test` the destination, see https://github.com/openshift/ci-tools/pull/3189#discussion_r1210419087)]
addSubgraph - post
    addNode - empty
```